app/src/main/java/com/arctouch/codechallenge/data/repository/Repository.kt

```diff
@@ -0,0 +1,41 @@
+package com.arctouch.codechallenge.data.repository
+
+import com.arctouch.codechallenge.data.model.GenreResponse
+import com.arctouch.codechallenge.data.model.UpcomingMoviesResponse
+import com.arctouch.codechallenge.data.service.TmdbApi
+import com.arctouch.codechallenge.util.BACKDROP_URL
+import com.arctouch.codechallenge.util.BASE_KEY
+import com.arctouch.codechallenge.util.POSTER_URL
+import io.reactivex.Single
+import io.reactivex.functions.BiFunction
+import java.util.*
+
+interface IRepository {
+
+    fun getUpcomingMovies(page: Long): Single<UpcomingMoviesResponse>
+}
+
+class Repository(private val tmdbApi: TmdbApi) : IRepository {
```

Comment:
Hello, about this i can do a implementation like this:

1. First i will implement the cache in the repository with a simple list

2. Create a sealed class like this:

`sealed class Result {
        object Loading : Result()
        data class Success(val movieList: List<Movie>) : Result()
        data class Failure(val throwable: Throwable) : Result()
    }`

3. In the method getUpcomingMovies i gonna change the return type for Result<List<Movie>>

`fun getUpcomingMovies(page: Long): Result<List<Movie>>`

4. After the process of the BiFunction to concat the results, i can implement a flatMap that convert the list to a Result<List<Movie>> and return. In case that contains some erro i treat this in the onErrorReturn, casting the error to Failure.

5. At the end i can use the method .startWith(Loading) in case if the cache is empty or .startWith(Result<Cache>) if is not.

6. In the viewModel i treat the result of the method:

`loading.set(result == Result.Loading)
        when (result) {
            is Result.Success -> { ... }
            is Result.Failure -> { ... }
       }
`

There is some parts of the implementation missing, but i think this pass my ideia, if this is not enought, please let me know i will try to explain in a different way.